openfisca_country_template/variables/housing.py

```diff
@@ -27,17 +27,9 @@ class rent(Variable):
     label = u"Rent paid by the household"
 
 
-#  Possible values for the housing_occupancy_status variable, defined further down
-class HousingOccupancyStatus(Enum):
-    tenant = u'Tenant'
-    owner = u'Owner'
-    free_lodger = u'Free logder'
-    homeless = u'Homeless'
-
-
 class housing_occupancy_status(Variable):
     value_type = Enum
-    possible_values = HousingOccupancyStatus
+    possible_values = HousingOccupancyStatus  # HousingOccupancyStatus is defined in base.py
```

Comment:
Why do we have to put all Enums in the base.py file ?
I know that is the way we did it for openfisca-France but I thought it was an idiosynchrasy of that package.

Reply:
I ran into a very stange bug when running core tests. After investigating, I found more deep issues:

[This import](https://github.com/openfisca/country-template/blob/master/openfisca_country_template/variables/taxes.py#L12) worked as expected by pure luck. It is not actually a relative import, but an absolute one. Relative imports are not allowed in our context, as the variables files are loaded one by one independently. It turns out the `housing.py` file was parsed first. Then the python importer recorded that it has loaded a module named `housing`. When it parses `taxes.py`, it sees `from housing import ...`. It then used the recorded module. If `taxes.py` was read before `housing.py`, the import would fail 😱.

To avoid this issue, the first thing I tried was to replace `from housing import HousingOccupancyStatus
` by `from openfisca_country_template.variables.housing import HousingOccupancyStatus`. With this the import works in all case, but another issue appears.

In the fast indexed enums, we rely on a [comparison of enums](https://github.com/openfisca/openfisca-core/blob/enums-perfs/openfisca_core/indexed_enums.py#L75) when comparing an `EnumArray` with an enum item. This only works if the two compared enums are _exactly_ the same object.

With the absolute import, we ended up importing `housing` twice: once when we parse `housing.py`, and another time when we parse `taxes.py` and run into the absolute import. Usually Python deals well with imports, but this doesn't work as we manually import modules from files. The double import make the comparison mentioned in the last paragraph to fail : the two enums compared have been imported separatly, they are not the same object.

Solutions i thought about:

- Make the enum comparison more robust. I didn't find an easy way to do it, and there is a risk to degrade the performances, and we know this part of the code must be fast.
- Fixing the way we import the variable files to avoid double imports. If we record the module`housing.py` as `openfisca_country_template.variables.housing` instead of just `housing`, the double import could be avoided. But I don't see any clean and robust way of doing it.
- Moving all the enums definitions to `base.py`. But it seems pretty heavy and not so  natural. I've reverted it.
- The new solution I implemented, which is avoiding any import by accessing the enum through the `EnumArray`. 

The last one seemed less heavy and more natural to me, but a different opinion on this is very much welcome !

Reply:
Sounds great, I will see if that change can be applied to openfisca-France :)